features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp

```diff
@@ -0,0 +1,469 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ble/gap/AdvertisingDataBuilder.h"
+
+namespace ble {
+
+AdvertisingDataBuilder::AdvertisingDataBuilder(mbed::Span<uint8_t> buffer) :
+    _buffer(buffer),
+    _payload_length(0) {
+}
+
+AdvertisingDataBuilder::AdvertisingDataBuilder(uint8_t* buffer, size_t buffer_size) :
+    _buffer(buffer, buffer_size),
+    _payload_length(0) {
+}
+
+mbed::Span<const uint8_t> AdvertisingDataBuilder::getAdvertisingData() const {
+    return _buffer.first(_payload_length);
+}
+
+ble_error_t AdvertisingDataBuilder::addData(
+    adv_data_type_t advDataType,
+    mbed::Span<const uint8_t> fieldData
+) {
+    if (findField(advDataType)) {
+        return BLE_ERROR_OPERATION_NOT_PERMITTED;
+    } else {
+        return addField(advDataType, fieldData);
+    }
+}
+
+ble_error_t AdvertisingDataBuilder::replaceData(
+    adv_data_type_t advDataType,
+    mbed::Span<const uint8_t> fieldData
+) {
+    uint8_t* field = findField(advDataType);
+
+    if (field) {
```

Comment:
Presumably this is doing a boolean test of the return value of a pointer , assuming that NULL pointer is false and any other pointer value is true. Bit hacky mixing of types. Pointer values should really be explicitly tested against NULL.

Reply:
It is more for clarity of types. Booleans should be treated as such and pointers should be treated as pointers. Even testing for a 0 value using a boolean operation isn't very nice, although perfectly legal

Reply:
Fixed.